### PR TITLE
RLQS: Refactor traffic processing in the RLQS filter & fix a broken expiration check

### PR DESCRIPTION
**Pull request by bsurber**
Commit Message: Refactor the RLQS filter's traffic processing & fix the broken action-assignment expiration check.
Additional Description:
Current behavior: 
- Every request that hits the RLQS filter results in the cache thinking that its entry has expired due to the initial / default action of a newly created BucketCache index not having a TTL.  
- If the TTL was fixed, it would still start failing all expiration checks shortly after assignments were received, even if those assignments were sent routinely (not going stale). This is because the expiration time was calculated off of a private field set on BucketCache index creation, and never updated when receiving assignments.  
- Note: Integration testing was previously passing as it was incorrectly expecting multiple RLQS usage reports.
Risk Level: minor - filter is in-development and currently in a broken state
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (tyxia) [inline source/extensions/filters/http/rate_limit_quota/filter.cc:335]: Is this true? if cached action expires, should we just fallback to expired assignment behavior?

File: source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h
Comment (tyxia) [inline source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h:45]: If default action is not necessary in this PR, i would suggest to do it in a separate PR.

Also, AFAIU when cache action expires, expired assignment behavior should be used rather than default action.

File: source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h
Comment (tyxia) [inline source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h:52]: nit: Do we really need this new variable? Can we just use `first_assignment_time`?

But it is up to you.

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (tyxia) [inline source/extensions/filters/http/rate_limit_quota/filter.cc:146]: From quick scan, IIUC. this update is the key. 

I am wondering if other changes (like default action changes) are needed for the test update?

**Review by tyxia — COMMENTED** [review]
Thank you! Left some comments from a quick first pass.

If possible, I would suggest having this PR only for fixing the expiration check and leaving the refactor into separate PR.

**Comment by tyxia** [comment]
Also please fix the format so that CI can run. You can use command `bazel run //tools/code_format:check_format -- fix`
Thanks!

/wait

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (tyxia) [reply source/extensions/filters/http/rate_limit_quota/filter.cc:335]: On second thought, I think your logic is as below, which make sense 
(a) first time If expired ---> 1. processExpiredBucket 2. reset `cached_action`
(b) then second time --->  2. use the default action

File: source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h:52]: If anything I'd delete `first_assignment_time` as we don't use that anywhere. It's not accurate to what we need to calculate expiration (thus the new variable).

Comment (tyxia) [reply source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h:52]: Yeah, if we actually only need one variable let's just keep one and rename it to make it accurate.

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/filter.cc:335]: If there's no cached action at all, it means that expiration has already happened in a previous run or no assignment has been received yet from the RLQS server. So yes, at this point, we're relying on the configured default.

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/filter.cc:146]: The `default_action` field is required to differentiate from the default-action set by the `no_assignment_behavior` field.
The `cached_action` is set by RLQS responses as they give assignments for buckets.

These are notably different in that `cached_action` has a TTL set by an RLQS response, whereas the `default_action` does not ever expire, and must always be available to handle traffic before an assignment is given or after the latest assignment has been wiped by expiration.

We could potentially remove the actual `default_action` field from the Bucket and rely on dynamically reading from the `no_assignment_behavior` field of the `match_action`, but this change more closely matches existing code that only reads from `no_assignment_behavior` during `createNewBucket`.

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (tyxia) [inline source/extensions/filters/http/rate_limit_quota/filter.cc:348]: Overall, I feel changing the allow/deny based on ret_status is not robust. The reason it works here is because our RLQS filter is in non-blocking fashion that it will not wait for response from RLQS server.

 `StopIteration` in Envoy generally represents that filter is waiting for the remote server response, rather than tightly tied with localReply/reject. 

Thus, can we update the quota_usage inline inside of the functions above when the request is determined to be ALLOW or Throttled.

**Review by tyxia — COMMENTED** [review]
Mostly LGTM modulo the open threads and code coverage  

Thanks!

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/filter.cc:145]: Added comments for clarity.

File: source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/quota_bucket_cache.h:53]: The var has been cleaned up.

File: source/extensions/filters/http/rate_limit_quota/filter.cc
Comment (bsurber) [reply source/extensions/filters/http/rate_limit_quota/filter.cc:340]: We've addressed this internally without too much to say. It'll do for now as that would be a bit of future-proofing for an unknown problem at the cost of additional refactoring.

**Review by tyxia — APPROVED** [review]
Thanks!

**Comment by phlax** [comment]
@tyxia @bsurber i think this is causing a very frequent flake - we may need to revert while the issue is resolved

**Comment by tyxia** [comment]
@phlax  Thanks for notification!

Could you please highlight which test is flaky? so that we can resolve/revert it accordingly

**Comment by phlax** [comment]
see eg here https://dev.azure.com/cncf/envoy/_build/results?buildId=178487&view=logs&j=1439b9f7-a348-5b50-b5fe-ea612ea91241&t=1002ac43-da84-5fae-70b2-98833b702d09&l=386

same test timed out twice on that run - but its happening a lot (if sporadically) elsewhere

**Comment by bsurber** [comment]
Hm, I had seen `ASSERT_TRUE(response_->waitForEndStream());` deadlock during testing but had thought I'd found the cause. Guess that was just occluded by flakiness and that search will have to resume.

**Pull request merged**
